src/DNA.cpp

```diff
@@ -489,3 +549,108 @@ string DNA::get_binary_representation(int value, int bit_count) {
 
     return output;
 }//end get_binary_representation
+
+/**
+ * rand_num
+ * 
+ * Returns a random integer [start, end).
+ **/
+int DNA::rand_num(int start, int end) {
+    int diff = end - start;
+    int r = rand() % diff;
+    r += start;
+    return r;
+}//end rand_num
+
+/**
+ * clip
+ * 
+ * Returns only the bits of value between [start, end). 
+ * Assumed that start <= end <= 64;
+ **/
+dna_t DNA::clip(dna_t value, int start, int end) {
+    long s = 0xffffffffffffffff << start;
+    long e;
+    if (end == 64 ) {
+        e = 0xffffffffffffffff;
+    } else {
+        e = 0x7fffffffffffffff >> (63 - end);
+    }
+    return (s & e) & value;
+}//end clip
+
+/**
+ * multipoint_crossover
+ * 
+ * Performs a multipoint crossover of the two dna 
+ * sequences with a given number of cuts. Cuts are 
+ * spread fairly evenly, but with some randomness.  
+ **/
+dna_t DNA::multipoint_crossover(dna_t a, dna_t b, int cut_num) {
+    //generate cross points:
+    vector<int> cuts(cut_num);
+    for (int i = 0; i < cut_num; i++) {
+        int start = floor(i * 64 / cut_num);
+        int end = floor((i+1) * 64 / cut_num);
+        cuts[i] = rand_num(start, end);
+    }
+
+    //random pick starting sequence:
+    dna_t current = (rand_num(0, 2)) ? a : b;
+
+    //splice new sequence:
+    dna_t new_sequence = 0;
+    int last_cut = 0;
+    for (int i = 0; i <= cuts.size(); i++) {
+        //determine end of clip:
+        int clip_end;
+        if (i == cuts.size()) {
+            clip_end = 64;
+        } else {
+            clip_end = cuts[i];
+        }
+
+        //get clip:
+        dna_t new_segment = clip(current, last_cut, clip_end);
+
+        //add clip to new sequence:
+        new_sequence = new_sequence | new_segment;
```

Comment:
Syntax question: Are you piping new_segment into new_sequence?? Because at first I thought it was the OR, but it's only one. Not familiar with that syntax.

Reply:
This is a bit-wise OR, not a logical OR. The sequence is all zeros in the segment which we are "inserting" the new segment to.